pkg/blockbuilder/scheduler/scheduler.go

```diff
@@ -78,24 +80,84 @@ func (s *BlockBuilderScheduler) running(ctx context.Context) error {
 
 func (s *BlockBuilderScheduler) updateSchedule(ctx context.Context) {
 	startTime := time.Now()
-	// Eventually this will also include job computation. But for now, collect partition data.
-	admin := kadm.NewClient(s.kafkaClient)
+	defer func() {
+		s.metrics.updateScheduleDuration.Observe(time.Since(startTime).Seconds())
+	}()
+
+	// TODO: Commit the offsets back to Kafka if dirty.
 
-	startOffsets, err := admin.ListStartOffsets(ctx, s.cfg.Kafka.Topic)
+	lag, err := blockbuilder.GetGroupLag(ctx, s.adminClient, s.cfg.Kafka.Topic, s.cfg.BuilderConsumerGroup, 0)
 	if err != nil {
-		level.Warn(s.logger).Log("msg", "failed to list start offsets", "err", err)
+		level.Warn(s.logger).Log("msg", "failed to get group lag", "err", err)
+		return
+	}
+
+	if ps, ok := lag[s.cfg.Kafka.Topic]; ok {
+		for part, gl := range ps {
+			partStr := fmt.Sprint(part)
+			s.metrics.partitionStartOffset.WithLabelValues(partStr).Set(float64(gl.Start.Offset))
+			s.metrics.partitionEndOffset.WithLabelValues(partStr).Set(float64(gl.End.Offset))
+			s.metrics.partitionCommittedOffset.WithLabelValues(partStr).Set(float64(gl.Commit.At))
+		}
 	}
-	endOffsets, err := admin.ListEndOffsets(ctx, s.cfg.Kafka.Topic)
+
+	oldTime := time.Now().Add(-s.cfg.ConsumeInterval)
+	oldOffsets, err := s.adminClient.ListOffsetsAfterMilli(ctx, oldTime.UnixMilli(), s.cfg.Kafka.Topic)
 	if err != nil {
-		level.Warn(s.logger).Log("msg", "failed to list end offsets", "err", err)
+		level.Warn(s.logger).Log("msg", "failed to obtain old offsets", "err", err)
+		return
 	}
 
-	s.metrics.updateScheduleDuration.Observe(time.Since(startTime).Seconds())
+	// See if the group-committed offset per partition is behind our "old" offsets.
 
-	startOffsets.Each(func(o kadm.ListedOffset) {
-		s.metrics.partitionStartOffset.WithLabelValues(fmt.Sprint(o.Partition)).Set(float64(o.Offset))
-	})
-	endOffsets.Each(func(o kadm.ListedOffset) {
-		s.metrics.partitionEndOffset.WithLabelValues(fmt.Sprint(o.Partition)).Set(float64(o.Offset))
+	oldOffsets.Each(func(o kadm.ListedOffset) {
+		if l, ok := lag.Lookup(o.Topic, o.Partition); ok {
+			if l.Commit.At < o.Offset {
+				level.Info(s.logger).Log("msg", "partition ready", "p", o.Partition)
+
+				// The job is uniquely identified by {topic, partition, consumption start offset}.
+				jobID := fmt.Sprintf("%s/%d/%d", o.Topic, o.Partition, l.Commit.At)
```

Comment:
With the aim of having only one job per partition at a time, I probably should not have the commit offset as part of the job ID.

Reply:
Thanks for this. Here's what I think:

The v1 guarantee of 1 job per partition (1JPP) is definitely something we want to uphold.

While I started with "topic/partition" IDs like you suggested, I've found including the offset makes a lot of sense because:
- It keeps logs about job progress/failure unambiguous and easy to correlate
- It simplifies our client communication model where clients repeatedly report historical job status
- It positions us well for future multi-job partitions

About the current code and how it guarantees 1JPP - yes, it's currently weak. It assumes offsets won't advance until the current job is complete. Given we have several services and multiple flavors of Kafka in play, that's not an assumption we should rely on.

Rather than using the jobQueue map key to enforce uniqueness of active partition consumption, I'm considering adding dedicated state in `BlockBuilderScheduler` to unambiguously track scheduled partitions. WDYT?